File: algorithms/labs/lab_6/D.java
/**
 * Nariman Safiulin (woofilee)
 * File: D.java
 * Created on: May 15, 2016
 */

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class D {
    public static final String PROBLEM_NAME = "pathbgep";

    public static void main(String[] args) throws Exception {
        Scanner in = new Scanner(new File(PROBLEM_NAME + ".in"));
        PrintWriter out = new PrintWriter(new File(PROBLEM_NAME + ".out"));

        solve(in, out);

        out.close();
        in.close();
    }

    private static class DirectedEdge implements Comparable<DirectedEdge> {
        int to;
        long weight;

        DirectedEdge(int to, long w) {
            this.to = to;
            this.weight = w;
        }

        @Override
        public int compareTo(DirectedEdge o) {
            return Long.compare(this.weight, o.weight);
        }
    }

    private static void solve(Scanner in, PrintWriter out) throws Exception {
        int n = in.nextInt();
        int m = in.nextInt();
        long[] distances = new long[n];
        ArrayList<DirectedEdge>[] edges = new ArrayList[n];

        for (int i = 0; i < n; i++) {
            edges[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            int f = in.nextInt() - 1;
            int s = in.nextInt() - 1;
            int w = in.nextInt();
            edges[f].add(new DirectedEdge(s, w));
            edges[s].add(new DirectedEdge(f, w));
        }

        Arrays.fill(distances, Long.MAX_VALUE);
        distances[0] = 0;
        PriorityQueue<DirectedEdge> q = new PriorityQueue<>();
        q.add(new DirectedEdge(0, 0));

        while (!q.isEmpty()) {
            int v = q.peek().to;
            if (q.poll().weight > distances[v])
                continue;

            for (DirectedEdge edge : edges[v]) {
                int to = edge.to;
                long w = edge.weight;
                if (distances[v] + w < distances[to]) {
                    distances[to] = distances[v] + w;
                    q.add(new DirectedEdge(to, distances[to]));
                }

            }
        }

        for (int i = 0; i < n; i++) {
            out.print(distances[i] + " ");
        }
    }

    static class Scanner {
        BufferedReader br;
        StringTokenizer t;

        Scanner(File file) throws Exception {
            br = new BufferedReader(new FileReader(file));
            t = new StringTokenizer("");
        }

        boolean hasNext() throws Exception {
            while (!t.hasMoreTokens()) {
                String line = br.readLine();
                if (line == null)
                    return false;
                t = new StringTokenizer(line);
            }
            return true;
        }

        String next() throws Exception {
            if (hasNext()) {
                return t.nextToken();
            }
            return null;
        }

        int nextInt() throws Exception {
            return Integer.parseInt(next());
        }

        long nextLong() throws Exception {
            return Long.parseLong(next());
        }

        double nextDouble() throws Exception {
            return Double.parseDouble(next());
        }

        void close() throws Exception {
            br.close();
        }
    }
}

File: algorithms/src/labs/lab_6/D.kt
/**
 * Nariman Safiulin (woofilee)
 * File: D.kt
 * Created on: May 15, 2016
 */

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.PrintWriter
import java.util.*

private val PROBLEM_NAME = "pathbgep"

private class Scanner(file: File) {
    val br = BufferedReader(FileReader(file))
    var st = StringTokenizer("")

    fun hasNext(): Boolean {
        while (!st.hasMoreTokens()) {
            val line = br.readLine() ?: return false
            st = StringTokenizer(line)
        }
        return true
    }

    fun next(): String? = if (hasNext()) st.nextToken() else null
    fun nextInt(): Int = Integer.parseInt(next())
    fun nextLong(): Long = java.lang.Long.parseLong(next())
    fun nextDouble(): Double = java.lang.Double.parseDouble(next())
    fun close() = br.close()
}

data class DirectedEdge(val to: Int, val weight: Long) : Comparable<DirectedEdge> {
    override fun compareTo(other: DirectedEdge): Int = when {
        this.weight - other.weight < 0 -> -1
        this.weight - other.weight > 0 -> 1
        else -> 0
    }
}

private fun solve(`in`: Scanner, out: PrintWriter) {
    val n = `in`.nextInt()
    val m = `in`.nextInt()
    val distances = Array(n) { Long.MAX_VALUE }
    val edges = Array(n) { ArrayList<DirectedEdge>(n) }

    (0..m - 1).forEach {
        val f = `in`.nextInt() - 1
        val s = `in`.nextInt() - 1
        val w = `in`.nextLong()
        edges[f].add(DirectedEdge(s, w))
        edges[s].add(DirectedEdge(f, w))
    }

    distances[0] = 0;
    val q = PriorityQueue<DirectedEdge>()
    q.add(DirectedEdge(0, 0))

    while (!q.isEmpty()) {
        val v = q.peek().to
        if (q.poll().weight > distances[v]) continue

        edges[v].forEach { edge ->
            val to = edge.to
            val w = edge.weight
            if (distances[v] + w < distances[to]) {
                distances[to] = distances[v] + w
                q.add(DirectedEdge(to, distances[to]))
            }

        }
    }

    (0..n - 1).forEach { i -> out.print("${distances[i]} ") }
}

fun main(args: Array<String>) {
    val `in` = Scanner(File(PROBLEM_NAME + ".in"))
    val out = PrintWriter(File(PROBLEM_NAME + ".out"))

    solve(`in`, out)

    `in`.close()
    out.close()
}
    